virtio-queue: update virtio-queue to v0.6.0

studychao opened the pull request:
Since the vm-memory conflict brought by virtio-queue 0.4.0, we need to upgrade virtio-queue to the newest version. Meanwhile, virtio-queue change a lot of APIs from 0.4.0 to 0.5.0, so this PR mainly for fixing all the API changes.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
virtio-queue: update virtio-queue to v0.6.0

Since the vm-memory conflict brought by virtio-queue 0.4.0, we need to
upgrade virtio-queue to the newest version. Meanwhile, virtio-queue
change a lot of APIs from 0.4.0 to 0.5.0, so this PR mainly for fixing
all the API changes.

fixes: #249
Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao commented on crates/dbs-virtio-devices/src/mmio/mmio_state.rs:
```diff
@@ -346,20 +346,32 @@ where
+
             for queue in self.queues.iter_mut() {
-                queue.queue = Q::new(queue.queue.max_size());
+                let new_queue = Q::new(queue.queue.max_size());
```
This part need to be reviewed more carefully.
Other parts are just changeing the parameter names.

studychao commented on crates/dbs-virtio-devices/src/device.rs:
```diff
@@ -68,8 +68,9 @@ impl<Q: QueueStateT> VirtioQueueConfig<Q> {
     pub fn create(queue_size: u16, index: u16) -> Result<Self> {
         let eventfd = EventFd::new(EFD_NONBLOCK).map_err(Error::IOError)?;
 
+        let queue = Q::new(queue_size)?;
```
This part need to be reviewed more carefully.
Other parts are just changeing the parameter names.

wllenyj commented on crates/dbs-virtio-devices/Cargo.toml:
```diff
@@ -32,7 +32,7 @@ serde_json = "1.0.9"
-virtio-queue = "0.4.0"
+virtio-queue = "0.6.0"
 vmm-sys-util = "0.11.0"
 vm-memory = { version = "0.9.0", features = [ "backend-mmap" ] }
```
Does this need to be updated too?

studychao replied:
I think there is no need. Since `virtio-queue` 0.6.0 use `vm-memory` 0.9.0 and 0.10.0 of vm-memory may cause conflict.

wllenyj commented on crates/dbs-virtio-devices/src/mmio/mmio_state.rs:
```diff
@@ -346,20 +346,32 @@ where
+                let new_queue = Q::new(queue.queue.max_size());
+                if let Err(e) = new_queue {
+                    reset_suc_flag = false;
+                    warn!("reset device failed because new virtio-queue could not be created due to {:?}", e);
```
If one of the Queues fails to initialize, the device does not work.

I think is it better to panic directly? @jiangliu

wllenyj replied:
Or modify the `reset` function to return `Result`, and the caller decides the next action.

studychao replied:
`Result` is added for `reset`.
And how the caller handle `reset` 's `Result` is also updated (in `mmio_v2.rs`).

studychao committed:
nydus: update nydus api, blobfs and rafs

In order to solve virtio-queue dependency problem, we need to update
nydus crates.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
crates: update ut related to virtio-queue

Update the unit tests that are related to virtio-queue upgrade changes.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
Clippy: fix clippy error

fix several clippy errors introduced by the latest Rust.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
ut: fix ut errors due to max queue size

adjust some queue size in UT to fix queue size erorr.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
nydus: update nydus related crates

update nydus-api, nydus-blobfs and nydus-rafs to fit with the
virtio-queue version.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

wllenyj commented on crates/dbs-virtio-devices/src/device.rs:
```diff
@@ -842,8 +842,8 @@ pub(crate) mod tests {
 
         let queues = vec![
-            VirtioQueueConfig::create(0, 0).unwrap(),
-            VirtioQueueConfig::create(0, 0).unwrap(),
```
What error will be reported if it is not modified?

studychao replied:
In the new Virtio-queue, when creating Virtio queue, it will check that the max size is a power of 2 because the valid queue sizes are a power of 2 as per the specification. 

So here we could not successfully create a Virtio queue with a max size of 0

wllenyj commented on crates/dbs-virtio-devices/src/fs/device.rs:
```diff
@@ -62,12 +61,6 @@ pub(crate) const PASSTHROUGHFS: &str = "passthroughfs";
 pub(crate) const RAFS: &str = "rafs";
 
-#[derive(Clone, Deserialize)]
-struct BlobCacheConfig {
```
Is this useless?

studychao replied:
Yes, in the new Nydus-api, they introduce `ConfigV2`, it could be created through `RafsConfig` str. `blob_config` is directly generated in the creation of `ConfigV2`, so there is no need for dbs-virtio-devices to define and create the BlobCacheConfig for Nydus.

wllenyj commented on crates/dbs-virtio-devices/src/vsock/muxer/muxer_impl.rs:
```diff
@@ -1,4 +1,5 @@
 // Copyright 2022 Alibaba Cloud. All Rights Reserved.
+// Copyright 2022 Alibaba Cloud. All Rights Reserved.
```
Repeated

studychao replied:
fixed

wllenyj commented on crates/dbs-virtio-devices/src/fs/device.rs:
```diff
@@ -295,10 +289,10 @@ impl<AS: GuestAddressSpace> VirtioFs<AS> {
                         "blob_cache_dir": "{}"
                     }}"#,
-                    cfg, source, &blob_config.work_dir
+                    cfg, source, blob_config.as_ref().unwrap().work_dir
```
Need to return Error here? This is the behavior at runtime.

wllenyj replied:
FsError::InvalidData

studychao replied:
fixed all the `unwrap()` with `ok_or()` to define `FsError`

wllenyj commented on crates/dbs-virtio-devices/src/fs/device.rs:
```diff
@@ -480,15 +474,14 @@ impl<AS: GuestAddressSpace> VirtioFs<AS> {
 
                 (
-                    Rafs::new(rafs_conf, mountpoint, &mut file)
+                    Rafs::new(&rafs_conf, mountpoint, &mut file)
```
Can it work without the reference?

studychao replied:
Nope, `Rafs` api has changed in the new Nydus.

wllenyj commented:
It seems that there are still a few clippy that have not been fixed.

studychao committed:
clippy: fix new clippy warnings introduced by rust 1.67

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>
